File: lecture4_jon/lecture4.go
package lecture4_jon

/*
* f(i) Objective function: distinct number way to reach the i-th of stair
* base cases:
  f(2) = 2
  f(1) = 1
  f(0) = 1 --> do nothing

  How to come to the top of the stair n, if we can just 1 or 2 steps?
  * it will be the sum of f(n - 1) + f(n - 2)
*/
func climb(n int) int {
	dp := make([]int, n+1)
	dp[0] = 1
	dp[1] = 1
	for i := 2; i <= n; i++ {
		dp[i] = dp[i-1] + dp[i-2]
	}
	return dp[n]
}

File: py/lecture4/lecture4.py
"""
Problem:
Find the number of way one can climb up a stair if he can climb 1 step or 2 steps at a time.

Objective function:
f(i) is the number of way one can climb up a stair

Recurrence relation:
f(n) = f(n-1) + f(n-2)
"""
def stair(n: int) -> int:
    dp = [0] * (n+1)
    dp[0] = 1
    dp[1] = 1
    for i in rang(2, len(dp)):
        dp[i] = dp[i-1] + dp[i-2]
    return dp[n]